feat: add ApplicationProfile listener

XDRAGON2002 opened the pull request:
Added a listener for `ApplicationProfile` which generates an `ApplicationProfile` for `pod-controller` based on given pod `ApplicationProfiles`.

NOTE: Not to be merged right now, more work is needed.

XDRAGON2002 commented:
@slashben Here is a listener which follows our discussion [here](https://cloud-native.slack.com/archives/C04GY6H082K/p1690802243973939).

I haven't been able to run/integrate this with the existing tool because apparently for some reason eBPF just doesn't want to work on a `WSL2 runtime` which is what I'm on right now.

Moreover a `DEVELOPMENT.md` would be extremely helpful as otherwise it would require contributors to read through `Makefile` and `scripts` folder.

Talking about the `scripts` folder, the scripts in there didn't run right off the bat for me, it was then resolved with [this](https://unix.stackexchange.com/questions/27054/bin-bash-no-such-file-or-directory/27067#27067).

Apart from these findings, just wanted to know if there was any way to get the project running on `WSL2` as otherwise I won't be able to integrate these changes to the code, even though only minor tweaks are needed, nevertheless even if there isin't a way to get this working on `WSL2`, at least I hope my code snippet is able to ease your/others work with this ;)

Thanks!

slashben commented:
> I haven't been able to run/integrate this with the existing tool because apparently for some reason eBPF just doesn't want to work on a `WSL2 runtime` which is what I'm on right now.

I see that people are battling WSL2 with eBPF. I saw two options that might help you ([this](https://gist.github.com/MarioHewardt/5759641727aae880b29c8f715ba4d30f) and [this](https://cloudblogs.microsoft.com/opensource/2022/02/22/getting-linux-based-ebpf-programs-to-run-with-ebpf-for-windows/).

Having said that, I think you can do a fair testing of your functionality without eBPF (!) with the way the `collector` and `eventsink` packages were constructed.

See [this test](https://github.com/slashben/kapprofiler/blob/80072795fdd47d400dc64549c3f534242d7f8703/pkg/collector/collector_test.go#L64) in [collector_test.go](https://github.com/slashben/kapprofiler/blob/main/pkg/collector/collector_test.go). The idea is that no real eBPF data is fed into the system, but `eventsink` is being fed by the test with mock data but still the test talks to the real K8s API.

I think you could write a test with the same "concept" where the test is creating and updating pod level ApplicationProfile object and your component does the aggregation. 

WDYT?


> Moreover a `DEVELOPMENT.md` would be extremely helpful as otherwise it would require contributors to read through `Makefile` and `scripts` folder.

Will do!

XDRAGON2002 commented:
Aha, I get it, instead of depending on eBPF, I'd just give mock data instead to check and see how it would behave in case it was run in an actual cluster.

Makes sense, I'll try this out.

XDRAGON2002 commented:
@slashben Voila! It works!

Here's what I did, made an extreme crude test that sets up all the necessary components and creates a deployment, I create the ApplicationProfiles of the pods of those deployments and deploy them in the cluster as well, I then go ahead and modify any one of the two and yep, it works!

(I know the process is extremely crude and can be made better, but this is not the final version so I'll improve upon this later down the line)

Now then what should the next step be? Currently the listener only acts up when an update occurs, I suppose we'd want to trigger it on creation and deletion of ApplicationProfiles as well? If so, I can go ahead and finish this up.

Also currently I've only added support for handling `Deployments` as `pod-controller` which all `pod-controllers` do we wish to support? `DaemonSet, StatefulSet`?

slashben commented:
Cool!

> Now then what should the next step be? Currently the listener only acts up when an update occurs, I suppose we'd want to trigger it on creation and deletion of ApplicationProfiles as well? If so, I can go ahead and finish this up.

Create is important for now, deletion is less, we can do it later

> Also currently I've only added support for handling `Deployments` as `pod-controller` which all `pod-controllers` do we wish to support? `DaemonSet, StatefulSet`?

Yes!

XDRAGON2002 commented:
@slashben I've added support for both, `creation` and `deletion` as well now.
Also added a check to ensure we only process `ApplicationProfiles` generated from `Pods`.
I've also ensured we are able to handle `Deployment`, `ReplicaSet`, `DaemonSet` and `StatefulSet` "pod-controllers".

Some improvements such as re-using the clients created in `main.go` and error handling are still not implemented, just wanted to get your thoughts as to what all needs to be done for this feature? If we're good here I can make the final changes and push them which includes the improvements I had mentioned earlier.

Thanks!

slashben commented on pkg/listener/listener.go:
```diff
@@ -0,0 +1,132 @@
+		controllerApplicationProfileRaw, _ := json.Marshal(controllerApplicationProfile)
+		_, _ = dynamicClient.Resource(collector.AppProfileGvr).Namespace(pod.Namespace).Patch(context.TODO(), podControllerName, apitypes.MergePatchType, controllerApplicationProfileRaw, metav1.PatchOptions{})
+	}
+}
```
What is unclear to me from the code is how the data is aggregated toward the owner reference.

I will write up a short proposal

slashben commented:
## Aggregation logic
The general idea here is the aggregate information from Pod-level application profiles based on owner references toward a higher-level workload construct.

Aggregation should happen every time an application profile at the Pod-level changes. In this event, the aggregation logic should look up the owner reference of the Pod (example: Daemonset), enumerate all the application profiles of the underlying Pods (with the same owner reference), and aggregate/merge the Pod-level application profiles. The output of this aggregation will be written back as an ApplicationProfile with the "kind/name" prefix of the owner on which the aggregation happened in the same namespace. 

Example: `daemonset-XXXXXX`

There has to be another aggregation level for the deployment level. The Deployments are owners of ReplicaSets and their aggregation should be different. In case an application profile of a ReplicaSet changes, the aggregator logic needs to look for the Deployment owner, and **copy** the ReplicaSet level application profile to the Deployment level application profile.

XDRAGON2002 commented:
Hi @slashben!

I agree with all the above mentioned edits, I'll make the appropriate changes.

Regarding how the aggregation occurs towards the owner references, lets take a glance at how I've implemented the flow.

We first check to see if the `ApplicationProfile` belongs to a `pod` or not, as otherwise it will lead to a loop trigger, change of `ApplicationProfile` of a `pod` will trigger change of the `pod controller ApplicationProfile` which will re-trigger the controller.

If the `ApplicationProfile` belongs to a `pod` we use a `switch` [here](https://github.com/slashben/kapprofiler/pull/1/files#diff-9d58ffc474c5e465149c2cb143824d4852cd800555a4869a3df8c3811dbf026bR68) to figure out who the owner of the pod is, if it's a `daemonset` or a `statefulset` we simply fetch them, get all their `pods` and aggregate their `ApplicationProfiles`.

In case the owner is a `replicaset` then there are two possibilities, either the `replicaset` was created by a `deployment` or was created directly by the user.

If it was created by the user we follow the same process as the `daemonset` and/or `statefulset`. Otherwise if it was created due to a `deployment`, we fetch that `deployment` and aggregate the results for that `deployment` as the owner of the pods completely bypassing the `replicaset` as the owner of the pods.

This is how the code is structured currently and looking at your above comment regarding the proposal I believe we are on the same page? 

Feel free to let me know if you'd want me to change the design, apart from that all the other changes regarding having a destroy function and using the same config defined in main and the merge logic are all things I had already planned for, so will get them done as soon as we finalize if the structure of the controller is correct :)

Thanks!

slashben commented:
> If it was created by the user we follow the same process as the `daemonset` and/or `statefulset`. Otherwise if it was created due to a `deployment`, we fetch that `deployment` and aggregate the results for that `deployment` as the owner of the pods completely bypassing the `replicaset` as the owner of the pods.

I have one reservation that needs consideration. The deployment is not equal to replicaset. If I have a deployment with image A and I decide to upgrade to image B, then Kubernetes will create a new replicaset based on the new image but the deployment will be the same. Since we are recording application behavior, it is important to bind our recording to a specific "application configuration" which is reflected by the image, the env variables and etc. Therefore aggregating/merging at the deployment level does not make sense. In every case, we need to aggregate at the replicaset level, then "point" from the deployment to the latest replicaset profile. My suggestion is to point by simply copying the latest replicaset level profile to deployment level profile.

> This is how the code is structured currently and looking at your above comment regarding the proposal I believe we are on the same page?

Except for the above yes :)

XDRAGON2002 commented:
@slashben I've made all the other required changes, just had a query regarding the "merge" logic, based on what field should the merge occur? 

For the `sysCalls` it's straightforward, just check for the string.

What about `execs` and `networkActivity`? As both are internally deeply nested structs.

Like for `execs` say for example, when are two of them considered equal? Should only the `path` be equal for both? Or does a deep compare need to occur to check that for two `execCalls` to be equal, they must have exact same `path`, `args` as well as `envs`?

Similarly what is the comparison process for `networkActivity` as it is much more densely nested then any of the others?

Apart from this, I've added tests and error handling as well. The above mentioned is probably the last piece of the puzzle before this can be merged ;)

Thanks!

slashben commented:
> For the `sysCalls` it's straightforward, just check for the string.

Agreed, it is just merging two string lists

 
> Like for `execs` say for example, when are two of them considered equal? Should only the `path` be equal for both? Or does a deep compare need to occur to check that for two `execCalls` to be equal, they must have exact same `path`, `args` as well as `envs`?

Two execs are considered equal when all they sub-fields are equal.

> Similarly what is the comparison process for `networkActivity` as it is much more densely nested then any of the others?

Same here. Two incoming or outgoing TCP connections are the same if the src/dest ip/port are the same.

In both cases, I suggest you define struct methods on the application profile items like 
```go
func (ap ApplicationProfileSpec) mergeInto(mergeAppProfile *ApplicationProfileSpec) error
func (cp ContainerProfile) mergeInto(mergeContainerProfile *ContainerProfile) error
func (cp ContainerProfile) mergeInto(mergeContainerProfile *ContainerProfile) error
func (ec ExecCalls) mergeInto(mergeExecCalls *ExecCalls) error
```

> Apart from this, I've added tests and error handling as well. The above mentioned is probably the last piece of the puzzle before this can be merged ;)

I saw you resolved some of my comments, but I haven't seen a new push fixing them. I guess the code is on your machine still :)

XDRAGON2002 commented:
@slashben I've pushed the changes, you can have a look now.

Let me if there are any further improvements required! Thanks!
